### DISCUSSION TITLE
Can artillery scenario runs a shell script

### DISCUSSION THREAD

**Discussion by bodycombathuang** [artilleryio/artillery#2322] [Artillery]
In https://www.artillery.io/docs/get-started/first-test example, in scenario part, we can call an HTTP URL  endpoint ,
does artillery support calling a shell script, I have a shell script which runs selenium automation , I want artillery to run the script, is it possbile? and how to? thanks

**Comment by bernardobridge** [comment]
Hi @bodycombathuang 👋 !

Artillery doesn't support calling a shell script out of the box, as it primarily lets you use YAML and NodeJS. However, you are able to extend Artillery by building your own engines and plugins to do pretty much whatever you need. So if you really needed this, you could follow [this guide](https://www.artillery.io/blog/extend-artillery-by-creating-your-own-engines) to build your own engine for running Selenium directly, or even for calling shell scripts (e.g. through using a library like `execa` or `shelljs`). Artillery's philosophy is to come with everything included for most use cases, and then allow you to extend it for custom use cases like yours.

I'd also point out that we do have direct out-of-the-box support for [browser automation through Playwright](https://www.artillery.io/docs/reference/engines/playwright). There's a lot of benefits to using that directly, as it's built-in to the tool, and it comes with a lot of features and optimisations to generate useful data during a test, as well as to run at scale (with Fargate support).

Hope that helps!

**Comment by bodycombathuang** [comment]
hi  @bernardobridge  is there any document on how to use shelljs and execa in artillery

**Comment by bodycombathuang** [comment]
hi @bernardobridge if I want to use shelljs and execa in artillery, I have to build my own engine?

**Reply by bernardobridge** [reply]
You can technically use it directly in Artillery functions (as long as the OS you're running on supports the libraries you're trying to use with `execa`/`shelljs`). 

**Example:**
the following will call the `callExecaScript` function (which you can then add your logic too), and emit a custom timer measuring how long it took executing it

```yaml
config:
   ...
   processor: 'PATH_TO_PROCESSOR_FUNCTION'

scenarios:
   - flow:
      - function: "callExecaScript"
```

```js
function callExecaScript(context, ee, next) {
    //Change your function name and add your logic here.
    //For more information, check: https://docs.art/http-reference#function-signatures

    const timeBefore = Date.now();
    //add custom logic here, including potentially calling execa/shelljs/anything else
    
    ee.emit('histogram', 'custom.selenium.time_taken', Date.now() - timeBefore)
    next();
};
```

---

Building your own `engine` just allows you to customise the interface much more to your needs, as well as control the custom metrics you create better. But it might be that the example above is enough for your needs.

**Comment by bodycombathuang** [comment]
hi @bernardobridge ,if I use
config:
   ...
   processor: 'PATH_TO_PROCESSOR_FUNCTION'

scenarios:
   - flow:
      - function: "callExecaScript"
      - 
     should I use engine in scenario? if I only want to call execascript

**Reply by bernardobridge** [reply]
You don't need to specify `engine`. It will use the `http` engine by default, but since that engine has support for running arbitrary functions during `flow`, you don't actually need to make HTTP calls during it.

**Comment by bodycombathuang** [comment]
I mean I do not want to call HTTP request and I do not want to call playwright

**Comment by bodycombathuang** [comment]
thanks let me try

**Comment by bodycombathuang** [comment]
hi @bernardobridge 
here is my javascript

```js
const execSync = require('child_process').execSync;
function callPlaywrightCommand(context, ee, next,tag,user,reportLocation) {
    return function(context,ee,next)
    {
        execSync(`runAutomation.bat ${tag} ${user} ${reportLocation}`);
        return next();
    }
}
module.exports = {
    callPlaywrightCommand: callPlaywrightCommand,
  };
```


this is my yml scenario

```yml
scenarios:
  - flow:
      - function: "callPlaywrightCommand('s5','user','hello')"
```
when I use artillery command to run the yml, it reports callPlaywrightCommand is undefined, why?>

**Reply by bernardobridge** [reply]
Hi @bodycombathuang . I've edited your post to use [Github Code Blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks) to make it more readable. If possible, please try to use that, as it makes it easier to read 🙇 .

---

The documentation for using functions is located here: https://www.artillery.io/docs/reference/engines/http#function-actions-and-beforescenario--afterscenario-hooks. You should take a look there for more information on how to use it.

The reason it's not working for you is because you're using the function in the YAML with arguments - this isn't currently possible. You'll have to do something like this:

```yml
scenarios:
  - flow:
      - function: "callPlaywrightCommand"
```

```js
const execSync = require('child_process').execSync;
function callPlaywrightCommandFactory(tag,user,reportLocation) {
    return function(context,ee,next)
    {
        execSync(`runAutomation.bat ${tag} ${user} ${reportLocation}`);
        return next();
    }
}

module.exports = {
    callPlaywrightCommand: callPlaywrightCommandFactory('s5', 'user', 'hello'),
  };
```

**Comment by bodycombathuang** [comment]
hi @bernardobridge , I do not want to pass ('s5', 'user', 'hello') in js file, 
is it possible I can  pass ('s5', 'user', 'hello') in yml?
why function: "callPlaywrightCommand('s5','user','hello')" does not work in yml, if callPlaywrightCommand('s5','user','hello') returns another function

**Reply by bernardobridge** [reply]
Hi @bodycombathuang 👋. No, as I explained above, it's not possible. The YAML file allows you to specify function names, and then those functions will be called with their function signature - no additional arguments possible. We may look to allow that in the future, but currently it's not something that has been implemented yet.

To have a dynamic element to it, you could try to use [variables](https://www.artillery.io/docs/reference/test-script#variables---inline-variables) in your test script. For example:

```yaml
config:
  variables:
    tag: "s5"
    user: "user"
    reportLocation: "hello"

scenarios:
  - flow:
      - function: "callPlaywrightCommand"
```

```js
const execSync = require('child_process').execSync;

function callPlaywrightCommand(context, ee, next) {
    execSync(`runAutomation.bat ${context.vars.tag} ${context.vars.user} ${context.vars.reportLocation}`);
        
    return next();
}

module.exports = {
    callPlaywrightCommand,
  };
```

---

To set some expectations, remember that you are trying to do something the tool isn't necessarily designed to do. I'm trying to provide a workaround with available features, so you don't have to build your own engine, but it will come with limitations :)

**Comment by bodycombathuang** [comment]
thank you
